crates/ruff_python_ast/src/name.rs

```diff
@@ -225,138 +349,457 @@ impl Display for UnqualifiedName<'_> {
+                }
+            }
+            SegmentsVec::Heap(heap) => heap.extend_from_slice(slice),
+        }
+    }
+}
+
+impl Default for SegmentsVec<'_> {
+    fn default() -> Self {
+        Self::new()
+    }
 }
 
-/// Convert an `Expr` to its [`QualifiedName`] segments (like `["typing", "List"]`).
-fn collect_segments(expr: &Expr) -> Option<SmallVec<[&str; 8]>> {
-    // Unroll the loop up to eight times, to match the maximum number of expected attributes.
-    // In practice, unrolling appears to give about a 4x speed-up on this hot path.
-    let attr1 = match expr {
-        Expr::Attribute(attr1) => attr1,
-        // Ex) `foo`
-        Expr::Name(nodes::ExprName { id, .. }) => {
-            return Some(SmallVec::from_slice(&[id.as_str()]))
+impl Debug for SegmentsVec<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_list().entries(self.as_slice()).finish()
+    }
+}
+
+impl<'a> Deref for SegmentsVec<'a> {
+    type Target = [&'a str];
+    fn deref(&self) -> &Self::Target {
+        self.as_slice()
+    }
+}
+
+impl<'a, 'b> PartialEq<SegmentsVec<'b>> for SegmentsVec<'a> {
+    fn eq(&self, other: &SegmentsVec<'b>) -> bool {
+        self.as_slice() == other.as_slice()
+    }
+}
+
+impl Eq for SegmentsVec<'_> {}
+
+impl Hash for SegmentsVec<'_> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_slice().hash(state);
+    }
+}
+
+impl<'a> FromIterator<&'a str> for SegmentsVec<'a> {
+    #[inline]
+    fn from_iter<T: IntoIterator<Item = &'a str>>(iter: T) -> Self {
+        let mut segments = SegmentsVec::default();
+        segments.extend(iter);
+        segments
+    }
+}
+
+impl<'a> From<[&'a str; 8]> for SegmentsVec<'a> {
+    #[inline]
+    fn from(segments: [&'a str; 8]) -> Self {
+        SegmentsVec::Stack(SegmentsStack {
+            segments,
+            len: segments.len(),
+        })
+    }
+}
+
+impl<'a> From<Vec<&'a str>> for SegmentsVec<'a> {
+    #[inline]
+    fn from(segments: Vec<&'a str>) -> Self {
+        SegmentsVec::Heap(segments)
+    }
+}
+
+impl<'a> Extend<&'a str> for SegmentsVec<'a> {
+    #[inline]
+    fn extend<T: IntoIterator<Item = &'a str>>(&mut self, iter: T) {
+        match self {
+            SegmentsVec::Stack(stack) => {
+                if let Err(segments) = stack.extend(iter) {
+                    *self = SegmentsVec::Heap(segments);
+                }
+            }
+            SegmentsVec::Heap(heap) => {
+                heap.extend(iter);
+            }
         }
-        _ => return None,
-    };
-
-    let attr2 = match attr1.value.as_ref() {
-        Expr::Attribute(attr2) => attr2,
-        // Ex) `foo.bar`
-        Expr::Name(nodes::ExprName { id, .. }) => {
-            return Some(SmallVec::from_slice(&[id.as_str(), attr1.attr.as_str()]))
+    }
+}
+
+const SMALL_LEN: usize = 8;
+
+#[derive(Debug, Clone, Default)]
+struct SegmentsStack<'a> {
+    segments: [&'a str; SMALL_LEN],
+    len: usize,
+}
```

Comment:
It's possible [`arrayvec`](https://docs.rs/arrayvec/latest/arrayvec/) could simplify things a bit here, although it [may have the same issue as `smallvec`](https://github.com/bluss/arrayvec/issues/96).

Reply:
Unfortunately, the `arrayvec` crate doesn't support the most complex operation, extending from an Iterator. That's why I think its use is limited. The main advantage I see is that it avoids initializing unused memory.

Reply:
I gave it a quick try. Funny enough, it has a variance issue, but it manifests differently than smallvec